### ISSUE TITLE
Tooling/functionality for recovering corrupted Containers with azure-client

### ISSUE THREAD

**Issue by skylerjokiel** [microsoft/FluidFramework#8870]
Developers need to be able to recover corrupted containers for their customers. This is a scenario more than a specific feature.

A proposed design is to expose a way for developers to get and load previous snapshots. This will allow customers to roll back in time to ensure not all the data is lost. While there is still data loss in this scenario it is minimized to a smaller window.

## Key Results
- [ ] Containers can be recovered with minimal data loss
- [ ] Developers understand how to use this functionality to recover from corruption scenarios.

## Epics
- [ ] #9442

**Comment by ssimic2** [comment]
@skylerjokiel could you drop off here if you know of any specific scenarios where we run into data corruption scenarios and subsequent actions. Or anything (anyone) pointing in that direction. I just wanted to get more background on practical use cases, what led to them, how we acted on them etc.

**Comment by ssimic2** [comment]
Various assumptions & considerations:

## Corrupted document – what is it?

If the latest snapshot or pending (not yet summarized) op is producing a state where document accuracy is compromised – we consider document to be “corrupted”. At that point, the client who determined corruption will close the container. This does not guarantee all connected clients will stop producing ops and summaries, thus often needing telemetry data to root cause the source of corruption.

## Causes of corruption

Current causes are rooted in FF actions/flows, and not directly tied up to user actions. Some examples:
-	Race conditions. (E.x Two users trying to create data store with the same name).
-	Bugs introduced with new op types.

## Learnings from ODSP docs recovery

After concluding that document is corrupted, the owner is advised to roll it back to some earlier version that was determined (by let’s say OCE) to be valid. The process is manually coordinated to make sure all clients have the doc closed before the recovery is kicked off. For ODSP docs, the recovery may involve going to SharePoint portal, and rolling the doc back to some earlier version. If all open instances are not closed during recovery, current summarizer may create another summary effectively bringing the document back to the original corrupted state.

OCE is normally able to point to:
1.	Exact op that introduced corrupted state. And/Or,
2.	Timestamp when corruption errors begun occurring.

## What can we do better?

- Have predictable recovery even if we have open instances of document.
- Better (more deterministic) handling of corrupted state.
- Make the rollback an atomic, streamlined process.

## Considerations

- Expose minimal ability needed to roll back versions - along the lines of “Roll back to version X” API. In other words, don’t allow users to upload their own summaries/snapshots or expose API surface that can lead to unintended uses.
- Consider authorization around who can initiate/orchestrate the rollback. This process will always be kicked off via explicit user actions.
- Consider offering API surface to expose state transitions during rollback. Host apps can leverage this API for proper user-messaging during and after recovery.
- Ensure we have sufficient snapshot history to minimize data loss.

## What we are NOT considering

- Improved ability to determine the root cause of corrupted state.
- Preventing corruption.
- Any UI admin panels for users to roll-back document versions. We are focusing only on API surface at this point.

**Comment by ssimic2** [comment]
Paths we are considering:

### Design path A: new container/doc hydrated from (earlier) snapshot of corrupted doc.

This route is very similar to hydrating from serialized summary of unattached container. The difference here is that we are using sanitized summary of another live container to hydrate newly created one. 

New API would encapsulate the following:
1. Fetch snapshot (belonging to corrupted container) at specific version.
2. Transform the snapshot to summary, while cleaning up any transient or unapplicable data (ex. Quorum).
3. Call `rehydrateDetachedFromSnapshot` to create and return new container.


Pros:
•	We are starting fresh, so we do not need to worry about the current state of the system (ops etc.) within corrupted container.
•	We do not need to worry about any connected clients and how they would go through the recovery process without affecting the outcome (ex. older summarizer creating another corrupted summary). We could still consider signaling corrupted container about presence of the recovered document, as an add-on improvement.
•	Other than the sanitization process and snapshot data transformation, we are relying on the existing capability of the system.

Cons:
•	We are creating a new document, with a new document ID. Client apps will need to be able to handle that.

### Design path B: Rolling back to an earlier snapshot version within existing document

New API would encapsulate the following:
1. Fetch snapshot (belonging to corrupted container) at specific version.
2. Transform the snapshot to summary, while cleaning up any transient or unapplicable data (ex. Quorum).
3. Signal open clients that recovery has begun. No further ops/summaries should be generated.
4. Force new summary upload (for the same container) out of data generated in step 2.
5. Signal open clients that recovery completed, and they can reload the container.

Pros:
•	We are recovering the original container. There is need to transition to another document during recovery.

Cons:
•	We would need signaling mechanism and state management logic to ensure correct behavior of all open clients during recovery.
•	“Discarded” ops are still present in the system as we do not have the capability to flush them. The implications (not yet investigated) may be a blocking factor here.
•	We may need to surface API for apps to communicate “recovery” state to the user, as they may be looking at the doc while it goes through recovery.
•	We need to expose capability to force summary uploads (step 4. above), that may be a backdoor for some unintended use cases.

**Comment by ssimic2** [comment]
Azure Fluid Relay API Gaps observed:

At minimum, we would rely on three API calls:
1.	Get snapshot history/versioning for a given container.
2.	Get snapshot/summary at a specific version.
3.	Upload summary.

Current gaps:

- Get snapshot/summary versioning history: timestamps and treeIDs seems to be off or missing.
- Get snapshot/summary at specific version: we always get only the “latest” summary.

**Comment by ssimic2** [comment]
@vladsud @agarwal-navin @anthony-murphy I'd love to get some feedback from runtime team on this.

**Comment by anthony-murphy** [comment]
For ODSP the framework doesn't do anything for recovery. ODSP has their versions, which can be rolled back to which the framework is agnostic to. These version roll backs, are also user initiated.  What the framework does support is a storage only mode, which is a read-only mode where the driver tells us not to even connect to the websocket. Odsp uses this to allow users to preview old versions from which they can copy data, or restore. OCE does not drive this process. When the rollback happens the documents, epoch will change, this is specific to the odsp driver. If the driver detects epoch change, it kills the session with a fatal error.

In general, this seems more a like a server feature than a client feature.

**Comment by vladsud** [comment]
Some corrections and note:
1. Some of the write up assumes access to a file (like reference that OCE knows op that caused trouble). That's not the case. Yes, in most cases we will not sequenceNumber, but not content of op. And we might not have access to a file (any non-MSFT files are out of reach)
2. OCE has no access to file or user identity (production ring). So, no OCE is involved here. User is initiating roll back.
3. There is no need to coordinate recovery (all clients closing container). The work to introduce checksums on ops would make it 100% reliable, but even with existing structure it's very hard to get some of the side-effects you mention. Basically, existing sessions may keep going for a while, but they will fail pretty quickly, and they should not impact file in storage.
   - I actually think epoch does not change in this process, but I'm not sure. If it does, then we have 100% correct system. If it does not - read on, but we have 99.99% correct system.
   - summary always references previous summary, so client that stays around after file was restored will not be able to summarize (unless restore got so lucky with restoring only couple ops, and only after reference point of new summary)
   - whenever client fetches new ops, we always do overlap fetches and always validate that file did not change by comparing storage ops with what client knows about
   - Ordering service (PUSH) will not be able to flush ops it has for active session and will close all connections. We need to clarify FRS story here
4. In general if you want to improve resiliency here, we need checksum work to complete, and I think it's stalled at the moment as nobody is putting cycles into it. Design is well understood, we just need someone to act on it.
5. Better (more deterministic) handling of corrupted state - not sure what you refer to. I'd think you refer to the fact that we might be not (eventually) consistent in how we handle failures. That depends on type of failure, so hard to say - most will be symmetrical for all clients, few will be bugs that may fail most clients but one.
6. Expose minimal ability needed to roll back versions - it would be nice to understand first why you think we need to support it at framework level, i.e., why it's not a functionality that is provided by storage (like in SPO case). Note that one would need to consider wider scenario - give user an ability first to see how document looks like using version X before rollback, and ability to undo it (return back to latest version before undo). While all of that can be done using existing storage APIs (i.e. framework can provide such capabilities), I'm not sure we should.
7. Save-As direction is great (snapshot -> new file), but this adds complexity for future (for example, SPO snapshots are not complete representation of file, there are resorces that are delay-loaded. Direction is fine, just mentioning that it has future cost for FRS (and not clear if this flow will work for 100GB+ file)

**Comment by ssimic2** [comment]
Great comments. Some new things I learned here. I did not know epoch tracker played a role in killing session.

Why client - In the short term, I was looking for lightweight solution that would be able to handle most common use cases, while still leveraging existing APIs. The solution that covers a smaller scope around “lets recover to the last known good version”, without necessarily requiring longer historical doc records (FRS cost increases with GB used), previewing ability, and other features we can certainly consider down the line (and outside the context of this problem). If the desire for a change/rollback was due to user actions, Id imagine we would have different reasoning here… To recover (with minimal data loss) we probably don’t even need the whole state (ops not yet summarized), or replaying ability, as last “valid” snapshot should be at some fairly recent version if we are able to quickly converge on “corrupted” state (and it seems like we can do that). Given some of these constraints and assumptions, looking at frameworks’ native capability seemed a good first step.

Good points about sizing, delay loading etc. We should have some reasonable expectations on what this API tooling needs to do, and what gaps (and when) need to be filled.

(@yangg-msft this is one of the topics of our earlier discussion.)

@vladsud can you clarify this: “whenever client fetches new ops, we always do overlap fetches and always validate that file did not change by comparing storage ops with what client knows about”. I’ll take a look at pending work you mentioned around checksums. But it sounds like (with 99.99% certainty) we can say that a) we do (fairly) quickly converge to “corrupted” state b) open clients cannot send us back to corrupted state, post-recovery.

“snapshot -> new file” is option A above. The downside of that option (depending how you look at it) is that we are creating a new document ID. We do not come out of recovery with rolled-back state, but with a new document. User needs to be able to handle that. Optionally, we could consider specifying ID during container creation (initial summary upload), and have some explicit “replace document” that server is able to handle. The downside is that we lose the history of the older (corrupted) doc. 

Is option B fully out of the question (using older snapshot to rehydrate existing/corrupted container) ? Even if we expose the ability to hydrate from `getSnapshpotTree` at specific older version (and run summary on demand) we will be messing up sequence numbering, older ops are still present (we don’t have ability to flush them) etc. I did not practically explore this, and I won’t go down this line if it’s clear this is not the path.

**Comment by vladsud** [comment]
@ssimic2 , please consider other areas. Whole data center will go down / restore to previous state, there will be fall overs with loss of tail of file, possibly storage provides way to restore files, or overwrite files. All of these scenarios will force client to reason about file identity and what to do when different content is in place of same file (or client is ahead and can't do much, but new clients start writing new history). Client needs to handle all of that, and that's way more important than corrupted containers IMHO.

Once there is handling for it, it feels natural that we do not build new mechanisms but rather rely on existing ones to determine it's a different file (and leverage same tools to reset identity of a file), and no matter who initiates recovery / file overwrite, it's just an overwrite and clients can recover same way as in above mentioned scenarios. Basically, storage should have capability to just overwrite a file. Client could indeed use snapshot -> summary conversion or simply re-summarize container at some seq# to recreate file.

Re overlapped reads - take a look at usage of comparableMessagePayload(), and code in fetchMissingDeltasCore() around from--; logic.

**Comment by ssimic2** [comment]
@yangg-msft please share details about computing/storage failover work. Let's see what functionality (if there) would make sense to expose for preview/recovery purposes. If we can leverage the mechanism there, great. If not, we'll proceed to explore these other options (short term options, leveraging existing APIs). What I have seen from OCE history, container corruption is legitimate issue - just devs inserting a bug (ex. faulty op) can have significant consequences. If that is not the case, lets re-evaluate the need for this ticket. Given the GA is here, we should look at high-priority items only.

@yangg-msft what is the timeline for FRS switching to full Git-based solution? This may open some optimization opportunities, if we end up considering one of the paths above.

**Comment by yangg-msft** [comment]
We're working on a document for the computing/storage failover. The doc should be ready later this week.

We expect to have a PoC for the Git-based summary storage available for you to experiment with towards the end of March, but the date for switching to it is TBD.

**Comment by ssimic2** [comment]
@yangg-msft if you have any drafts (computing/storage failover) that can just help us evaluate mechanism in the context of being the possible solution for this problem (recovering corrupted containers), please share those.

As for your earlier question, my assumption is that "file" is really just snapshot + ops.

**Comment by ssimic2** [comment]
Closing this in favor of #9442.

**Comment by skylerjokiel** [comment]
Re-opening to serve as Story
